sql: Adding support for int2, int8, float4 and float8 to be in line with postgres support of numerical types.

m-schneider opened the pull request:
Before this change we didn't support any of the types mentioned above.

After this change we'll support those types as alias and we will check
for overflow on INSERT and UPDATE. We currently won't support type checking
on mathematical operations, inline with how decimals are currently implemented.
Support for those operations would require a change to Datums, which is outside
of the scope for this change.

Closes #12481
Closes #14493

Commit:
sql: Adding support for int2, int8, float4 and float8 to be in line
with postgres support of numerical types.

Before this change we didn't support any of the types mentioned above.

After this change we'll support those types as alias and we will check
for overflow on INSERT and UPDATE. We currently won't support type checking
on mathematical operations, inline with how decimals are currently implemented.
Support for those operations would require a change to Datums, which is outside
of the scope for this change.

Closes #12481
Closes #14493

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16720)
<!-- Reviewable:end -->

knz commented:
Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/table, line 309 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd7Zzde6mxq5xRaOOn:-Knd7ZzeMZzX2v-KFkcc:b-6g6307) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/logictest/testdata/logic_test/table#L309)):*
> ```
> statement ok
> CREATE TABLE test.alltypes (
>   a BOOL,
> ```

I'm really sorry that you had to perform this renumbering of the variables, and I am sorry in advance for the sentence that follows,

but please do it just once more, so that *nobody else will have to do it again in the future*: rename all the columns with the form `c`+type, for example `cbool`, `cint`, `cint2` etc.
You can automate a lot of this with a regular expression for example `s/([^ ]*) ([^,]*),/c\2 \2/g`

Then sort the definitions in alphabetical order.

---

*[pkg/sql/parser/col_types.go, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd8K8g0RWFfpCkiyZW:-Knd8K8g0RWFfpCkiyZX:bux07pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L97)):*
> ```Go
> // Pre-allocated immutable integer column types.
> var (
> 	intColTypeBit         = &IntColType{Name: "BIT", N: 1, ImplicitWidth: true}
> ```

Huh I now see this parameter "N". Why not reuse this for int types throughout instead of "precision"?
Or perhaps also use "precision" for type `BIT`.

---

*[pkg/sql/parser/col_types.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd94HBH8aL1p5eaaUB:-Knd94HCqbBQeQd8KuDp:bslhm3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L124)):*
> ```Go
> 	Name          string
> 	N             int
> 	Precision     int
> ```

Yeah I think what "N" does is exactly what you need here, no need for a new field.

Also as revealed in package `sqlbase` the right name for this field is neither "N" or "Precision" but "Width".
Currently we have there `col.Type.Width = int32(t.N)` this would be best written as `col.Type.Width = int32(t.Width)`

Note that this comment is compatible with the concepts in postgres: in pg only types numeric and float have a "precision"; the integer types have a "size".

---

*[pkg/sql/parser/col_types.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndA0RteUY1pdNNA4k0:-KndA0RteUY1pdNNA4k1:b-huyq8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L147)):*
> ```Go
> 	floatColTypeReal   = &FloatColType{Name: "REAL"}
> 	floatColTypeFloat  = &FloatColType{Name: "FLOAT"}
> 	floatColTypeFloat4 = &FloatColType{Name: "FLOAT4", Prec: 4, PrecSpecified: true}
> ```

There's a conceptual mistake here: 
FLOAT4 and FLOAT8 differ in *width* (single vs double precision) - see https://www.postgresql.org/docs/9.6/static/datatype.html#DATATYPE-TABLE
FLOAT(4) and FLOAT(8) are *both single precision*, whereas double precision is selected with FLOAT(25) to FLOAT(54) - see https://www.postgresql.org/docs/9.6/static/datatype-numeric.html#DATATYPE-FLOAT

| Type name | Width (conceptual storage size, also determines max exponent) | Guaranteed result precision (in bits, determines size of mantissa) | Notes |
|------|------------|------------------|----------|
| real | 32 | 24 | we don't support this correctly yet but we should |
| float | 64 | 53 | |
| float4 | 32 | 24 | synonym to `real` |
| float8 | 64 | 24 | synonym to `float` |
| float(4) | 32 | 4 |
| float(8) | 32 | 8 |
| float(24) | 32 | 24 |
| float(25) | 64 | 25 |


So you need two separate fields "width" and "prec" in FloatColType and handle them properly here, in Format and in computations.

---

*[pkg/sql/sqlbase/table.go, line 1641 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndBzDu3fF0CqOyNYpE:-KndBzDu3fF0CqOyNYpF:b3rkrl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/sqlbase/table.go#L1641)):*
> ```Go
> 				}
> 			}
> 			if col.Type.Precision > 0 {
> ```

look 5 lines above this is equivalent to what's done with the field `width`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/col_types.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndA0RteUY1pdNNA4k0:-KndCs0Ooy4JrsXKB5GK:b-5l17kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

There's a conceptual mistake here: 
FLOAT4 and FLOAT8 differ in *width* (single vs double precision) - see https://www.postgresql.org/docs/9.6/static/datatype.html#DATATYPE-TABLE
FLOAT(4) and FLOAT(8) are *both single precision*, whereas double precision is selected with FLOAT(25) to FLOAT(54) - see https://www.postgresql.org/docs/9.6/static/datatype-numeric.html#DATATYPE-FLOAT

| Type name | Width (conceptual storage size, also determines max exponent) | Guaranteed result precision (in bits, determines size of mantissa) | Notes |
|------|------------|------------------|----------|
| real | 32 | 24 | we don't support this correctly yet but we should |
| float | 64 | 53 | |
| float4 | 32 | 24 | synonym to `real` |
| float8 | 64 | 24 | synonym to `float` |
| float(4) | 32 | 4 |
| float(8) | 32 | 8 |
| float(24) | 32 | 24 |
| float(25) | 64 | 25 |


So you need two separate fields "width" and "prec" in FloatColType and handle them properly here, in Format and in computations.
</blockquote></details>

note that in computations perhaps right now we don't need to do anything, because we do not have separate DFloat32 and DFloat64. But the correct way would be to have both and use the width field to select one or another (https://stackoverflow.com/questions/16889042/postgresql-what-is-the-difference-between-float1-and-float24#16889190)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 7 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/table, line 309 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd7Zzde6mxq5xRaOOn:-KneBwU7GC5N6dYKM1iI:b-18rpz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/logictest/testdata/logic_test/table#L309)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm really sorry that you had to perform this renumbering of the variables, and I am sorry in advance for the sentence that follows,

but please do it just once more, so that *nobody else will have to do it again in the future*: rename all the columns with the form `c`+type, for example `cbool`, `cint`, `cint2` etc.
You can automate a lot of this with a regular expression for example `s/([^ ]*) ([^,]*),/c\2 \2/g`

Then sort the definitions in alphabetical order.
</blockquote></details>

Column Selection mode FTW!

---

*[pkg/sql/parser/col_types.go, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd8K8g0RWFfpCkiyZW:-Kne7Paauvqf74cgD06K:bimv512) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L97)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Huh I now see this parameter "N". Why not reuse this for int types throughout instead of "precision"?
Or perhaps also use "precision" for type `BIT`.
</blockquote></details>

I changed both to width.

---

*[pkg/sql/parser/col_types.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd94HBH8aL1p5eaaUB:-Kne7VVyk00QXAZ35uVP:bk9zxen) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L124)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yeah I think what "N" does is exactly what you need here, no need for a new field.

Also as revealed in package `sqlbase` the right name for this field is neither "N" or "Precision" but "Width".
Currently we have there `col.Type.Width = int32(t.N)` this would be best written as `col.Type.Width = int32(t.Width)`

Note that this comment is compatible with the concepts in postgres: in pg only types numeric and float have a "precision"; the integer types have a "size".
</blockquote></details>

Changed both to width.

---

*[pkg/sql/sqlbase/table.go, line 1641 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndBzDu3fF0CqOyNYpE:-KneFaCIh1ImaOYfTs88:bu39jz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/sqlbase/table.go#L1641)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

look 5 lines above this is equivalent to what's done with the field `width`
</blockquote></details>

I see what you mean about bit strings, but the definition for integer types is here:
https://www.postgresql.org/docs/9.4/static/datatype-numeric.html#DATATYPE-INT and this function is inline with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/table, line 309 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knd7Zzde6mxq5xRaOOn:-KneGxmSLb5SqYwC8ao_:b5tot1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/logictest/testdata/logic_test/table#L309)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Column Selection mode FTW!
</blockquote></details>

:+1:
Thank you!

---

*[pkg/sql/sqlbase/table.go, line 1641 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndBzDu3fF0CqOyNYpE:-KneHTdH86BAwVgUsmgp:bb7uoub) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/sqlbase/table.go#L1641)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I see what you mean about bit strings, but the definition for integer types is here:
https://www.postgresql.org/docs/9.4/static/datatype-numeric.html#DATATYPE-INT and this function is inline with that.
</blockquote></details>

1) actually I hadn't seen, but the math below is ... inelegant. There's a poor trade-off here: the simple check above would be much easier on the eye and the mind of future maintainers than trying to comprehend your formula below, considering it must be first extended  with and extensive comment in which you would need to explain why the math happens to be correct, although there are many reasons why it may seem wrong at first sight.
2) you can extend the math above to work for both BIT (an unsigned type) and INT(N) (a signed type, where either one of two loops must be used to compute the magnitude depending on the sign). Then you can replace the message `bit string too long for type %s (column %q)` by `integer out of range for type %s (column %q)`, and that would be correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Just for the sake of clarity, some of the many reasons why the math appears wrong:
- it's quite bad to invoke a FP function (`math.Pow`) for something that should be doable in pure integer arithmetic
- int -> float conversion is inexact for numbers larger than 2^53, so it's not clear why the formula works with a width field value larger than 53

m-schneider commented:
FP function removed.

---

Review status: 9 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/table.go, line 1641 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndBzDu3fF0CqOyNYpE:-Knefq6ChVSspnAgbPNh:b41myiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/sqlbase/table.go#L1641)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) actually I hadn't seen, but the math below is ... inelegant. There's a poor trade-off here: the simple check above would be much easier on the eye and the mind of future maintainers than trying to comprehend your formula below, considering it must be first extended  with and extensive comment in which you would need to explain why the math happens to be correct, although there are many reasons why it may seem wrong at first sight.
2) you can extend the math above to work for both BIT (an unsigned type) and INT(N) (a signed type, where either one of two loops must be used to compute the magnitude depending on the sign). Then you can replace the message `bit string too long for type %s (column %q)` by `integer out of range for type %s (column %q)`, and that would be correct.
</blockquote></details>

I simplified the code for both bits and integers. I don't quite understand the need for the loop for bits. Using a loop for negative numbers would just be incorrect because Go uses 2s compliment to store negative numbers so our internal int64 representation would overflow for any negative number by definition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KneWWAEUIMpUN7yMM8x:bnmzq9x)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/col_types.go, line 171 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfW46oHuIhZiPg6XyG:-KnfW46pG3H0nvdrCTHq:b-719eo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types.go#L171)):*
> ```Go
> func (node *FloatColType) Format(buf *bytes.Buffer, f FmtFlags) {
> 	buf.WriteString(node.Name)
> 	if node.Prec > 0 && node.Width == 0 {
> ```

I think you should remove the second part of the conditional.
`FLOAT4(10)` is actually sensical and perhaps useful.

---

*[pkg/sql/parser/col_types_test.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVcP_dyzV7tqiPyWI:-KnfVcPaJ6mfCP0hvgRt:b-vdclpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L45)):*
> ```Go
> 		{"DOUBLE PRECISION", &FloatColType{Name: "DOUBLE PRECISION"}},
> 		{"FLOAT", &FloatColType{Name: "FLOAT"}},
> 		{"FLOAT4", &FloatColType{Name: "FLOAT4", Prec: 24, Width: 16, PrecSpecified: true}},
> ```

Drop "Prec" (using the default here)
Width:32

---

*[pkg/sql/parser/col_types_test.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVdCwTzRKkrcXTmfj:-KnfVdCx4oVX3VESlx9v:b914xhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L46)):*
> ```Go
> 		{"FLOAT", &FloatColType{Name: "FLOAT"}},
> 		{"FLOAT4", &FloatColType{Name: "FLOAT4", Prec: 24, Width: 16, PrecSpecified: true}},
> 		{"FLOAT8", &FloatColType{Name: "FLOAT8", Prec: 24, Width: 32, PrecSpecified: true}},
> ```

Drop "Prec" (using the default)
Width: 64

---

*[pkg/sql/parser/col_types_test.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfWM6F6w9HonvGoYh0:-KnfWM6F6w9HonvGoYh1:b-rogd83) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L47)):*
> ```Go
> 		{"FLOAT4", &FloatColType{Name: "FLOAT4", Prec: 24, Width: 16, PrecSpecified: true}},
> 		{"FLOAT8", &FloatColType{Name: "FLOAT8", Prec: 24, Width: 32, PrecSpecified: true}},
> 		{"FLOAT(4)", &FloatColType{Name: "FLOAT", Prec: 4, PrecSpecified: true}},
> ```

Add a test for FLOAT4(10) and FLOAT8(30)

---

*[pkg/sql/sqlbase/table.go, line 1641 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndBzDu3fF0CqOyNYpE:-KnfZfg2yACroRE9HfBQ:b-g9e3kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/sqlbase/table.go#L1641)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I simplified the code for both bits and integers. I don't quite understand the need for the loop for bits. Using a loop for negative numbers would just be incorrect because Go uses 2s compliment to store negative numbers so our internal int64 representation would overflow for any negative number by definition.
</blockquote></details>

Yes you are right. See my other comment for the followup.

---

*[pkg/sql/sqlbase/table.go, line 1621 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfX9noHKLHa3PhLyCv:-KnfX9npAEZsy-T5ylkM:b-u6wjaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/sqlbase/table.go#L1621)):*
> ```Go
> 		if v, ok := parser.AsDInt(val); ok {
> 			if col.Type.Width > 0 {
> 				if col.Type.VisibleType == ColumnType_NONE {
> ```

Scratch this entire conditional and its `else` branch. Also beware in the code you wrote if `Type.Width = 64` then `width = 63` then `1<<width=(-2^63)` so `1<<width-1` is undefined, so your formula is still tricky to get right. I think this works:

```go
if col.Type.Width > 0 {
   shifted := v >> uint(col.Type.Width-1)
   if (v >= 0 && shifted != 0) || (v < 0 && shifted != -1) {
       return fmt.Errorf("integer out of range for type %s (column %q)", col.Type.VisibleType, col.Name)
   }
}
```

Add tests for the boundary values: all combinations of Width = 0, 1, 63, 64, values = 0, 1, -1, MaxInt64, MinInt64.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 5 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/parser/col_types.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndA0RteUY1pdNNA4k0:-KnjHLFCi47xhTUzvpAO:be9lfr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

note that in computations perhaps right now we don't need to do anything, because we do not have separate DFloat32 and DFloat64. But the correct way would be to have both and use the width field to select one or another (https://stackoverflow.com/questions/16889042/postgresql-what-is-the-difference-between-float1-and-float24#16889190)
</blockquote></details>

So should we do anything extra in this PR to support all of this correctly, or wait for the change at the datum level when we support float32 and float64.

---

*[pkg/sql/parser/col_types.go, line 171 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfW46oHuIhZiPg6XyG:-KnjH490mY7MQ7_AZ-DV:bss4a1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types.go#L171)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think you should remove the second part of the conditional.
`FLOAT4(10)` is actually sensical and perhaps useful.
</blockquote></details>

Removed!

---

*[pkg/sql/parser/col_types_test.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVcP_dyzV7tqiPyWI:-KnjH7PiMCcBLi1D6l-9:bnvutbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L45)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Drop "Prec" (using the default here)
Width:32
</blockquote></details>

Dropped

---

*[pkg/sql/parser/col_types_test.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVdCwTzRKkrcXTmfj:-KnjHizkJJCZDDbJZL5Z:bu4ephv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L46)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Drop "Prec" (using the default)
Width: 64
</blockquote></details>

Dropped.

---

*[pkg/sql/parser/col_types_test.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfWM6F6w9HonvGoYh0:-KnjHAydEFkLCAlRvctB:b2pel6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L47)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Add a test for FLOAT4(10) and FLOAT8(30)
</blockquote></details>

I'm not convinced that we should add support for that. As of now those two string won't be parsed by sql.go.

---

*[pkg/sql/sqlbase/table.go, line 1621 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfX9noHKLHa3PhLyCv:-KnjI2B5cfrPN-thc_dj:buyn2rt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/sqlbase/table.go#L1621)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Scratch this entire conditional and its `else` branch. Also beware in the code you wrote if `Type.Width = 64` then `width = 63` then `1<<width=(-2^63)` so `1<<width-1` is undefined, so your formula is still tricky to get right. I think this works:

```go
if col.Type.Width > 0 {
   shifted := v >> uint(col.Type.Width-1)
   if (v >= 0 && shifted != 0) || (v < 0 && shifted != -1) {
       return fmt.Errorf("integer out of range for type %s (column %q)", col.Type.VisibleType, col.Name)
   }
}
```

Add tests for the boundary values: all combinations of Width = 0, 1, 63, 64, values = 0, 1, -1, MaxInt64, MinInt64.
</blockquote></details>

I have those tests in Scale already for int2, int4 and the bit strings.  I also needed to add another check for BITs specifically because they're unsigned so using the same check would fail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/col_types.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndA0RteUY1pdNNA4k0:-Knjp5u28dGah9ksMkC3:b-qohk7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L147)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

So should we do anything extra in this PR to support all of this correctly, or wait for the change at the datum level when we support float32 and float64.
</blockquote></details>

Up to you. However I do still believe the two types here should be initialized with Width 32 and 64 respectively not 16 and 32.

---

*[pkg/sql/parser/col_types_test.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVcP_dyzV7tqiPyWI:-KnjpG_IIGLb56tmE4BK:bhrnhhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L45)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Dropped
</blockquote></details>

Width: 32

---

*[pkg/sql/parser/col_types_test.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVdCwTzRKkrcXTmfj:-KnjpIH8R5Bbh7QUXHAI:bhrnhk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L46)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Dropped.
</blockquote></details>

Width: 64

---

*[pkg/sql/parser/col_types_test.go, line 47 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfWM6F6w9HonvGoYh0:-KnjpJmXUOh-rcQANID5:b6dsug9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L47)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I'm not convinced that we should add support for that. As of now those two string won't be parsed by sql.go.
</blockquote></details>

Ack. I'm good with that.

---

*[pkg/sql/sqlbase/table.go, line 1621 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfX9noHKLHa3PhLyCv:-KnjpaojLDyatIfD3cu1:be17u80) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/sqlbase/table.go#L1621)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I have those tests in Scale already for int2, int4 and the bit strings.  I also needed to add another check for BITs specifically because they're unsigned so using the same check would fail.
</blockquote></details>

Ack, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 10 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/parser/col_types.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KndA0RteUY1pdNNA4k0:-Knjs4Q0TdxC5m9nUVjj:bt2vr0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dd8d0f45d67644937d2e7cc7893d07f96d23376/pkg/sql/parser/col_types.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Up to you. However I do still believe the two types here should be initialized with Width 32 and 64 respectively not 16 and 32.
</blockquote></details>

Fixed!

---

*[pkg/sql/parser/col_types_test.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVcP_dyzV7tqiPyWI:-KnjrkyMRe2jK_PzAxq4:b-kbmdys) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L45)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Width: 32
</blockquote></details>

That's right thanks!

---

*[pkg/sql/parser/col_types_test.go, line 46 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnfVdCwTzRKkrcXTmfj:-Knjrn36dDWgfkVVPCgW:b-896gsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6adb980f9ff7c3777f826ab72eaa148b13672b8/pkg/sql/parser/col_types_test.go#L46)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Width: 64
</blockquote></details>

Don.e

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/col_types.go, line 144 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knjt6SkqCU5E5G78OPe:-Knjt6SkqCU5E5G78OPf:bnm0wr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d17d4a621f7a93db0f0045e7b4f71b61d345ba/pkg/sql/parser/col_types.go#L144)):*
> ```Go
> // Pre-allocated immutable float column types.
> var (
> 	floatColTypeReal   = &FloatColType{Name: "REAL"}
> ```

Why not populate `Width` properly for every FLOAT type?

---

*[pkg/sql/sqlbase/structured.go, line 1722 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnjtOQklu3RCkO-5Sjs:-KnjtOQklu3RCkO-5Sjt:b-wx3eye) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d17d4a621f7a93db0f0045e7b4f71b61d345ba/pkg/sql/sqlbase/structured.go#L1722)):*
> ```Go
> 		}
> 	case ColumnType_FLOAT:
> 		if c.Precision > 0 && c.VisibleType == ColumnType_NONE {
> ```

I still think you do not need the 2nd part of this conditional, even if you decide to not support the corresponding input syntax.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 8 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/col_types.go, line 144 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knjt6SkqCU5E5G78OPe:-KnjvXDuJUBPuTihEzMq:b-80tlqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d17d4a621f7a93db0f0045e7b4f71b61d345ba/pkg/sql/parser/col_types.go#L144)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why not populate `Width` properly for every FLOAT type?
</blockquote></details>

Sure.

---

*[pkg/sql/sqlbase/structured.go, line 1722 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnjtOQklu3RCkO-5Sjs:-KnjvNb7uf3mdbQlOqCE:btdaj7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3d17d4a621f7a93db0f0045e7b4f71b61d345ba/pkg/sql/sqlbase/structured.go#L1722)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I still think you do not need the 2nd part of this conditional, even if you decide to not support the corresponding input syntax.
</blockquote></details>

Agreed, this is before I was using default precision for the new types.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-Knjyy7Yabn9VnX_LXC9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Nice work so far! There is a little bit more to be done to make this complete, however.

The missing feature is related to type identifiers on pgwire. With this patch, even though we can create tables with e.g. `INT2`columns, the type of the column will still be reported to Postgres wire clients as `int8`. Additionally, the metadata reflected in `pg_catalog.pg_attribute` will be incorrectly reported as `int8`.

To demonstrate this problem, observe the difference between the following queries:

CockroachDB:

```
root@:26257/test> CREATE TABLE types(a int8, b int2);
CREATE TABLE
root@:26257/test> SELECT attname, atttypid, typname FROM pg_attribute a JOIN pg_type t ON a.atttypid=t.oid WHERE attrelid = 'types'::REGCLASS;
+---------+----------+---------+
| attname | atttypid | typname |
+---------+----------+---------+
| a       |       20 | int8    |
| b       |       20 | int8    |
+---------+----------+---------+
(2 rows)
```

Postgres:
```
jordan=# CREATE TABLE types(a int8, b int2);
CREATE TABLE
jordan=# SELECT attname, atttypid, typname FROM pg_attribute a JOIN pg_type t ON a.atttypid=t.oid WHERE attrelid = 'types'::REGCLASS;
 attname  | atttypid | typname
----------+----------+---------
 a        |       20 | int8
 b        |       21 | int2
 tableoid |       26 | oid
 ctid     |       27 | tid
 xmin     |       28 | xid
 xmax     |       28 | xid
 cmin     |       29 | cid
 cmax     |       29 | cid
(8 rows)
```

(Ignore the bottom 6 rows, those are Postgres internals that we don't care about).

You can see the difference - Postgres reflects a different `atttypid` in `pg_attribute` depending on the type of the column. We report both columns as an int8.

To fix this we'll need to propagate the additional column type information that you added up into the SQL type defined in `sql/type.go`. The main entry point to that change will be `ToDatumType` in `structured.go`.

---

Review status: 12 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/scale, line 61 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk2RUEKccrcHEscNJd:-Knk2RUEKccrcHEscNJe:b-3u3jtb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/logictest/testdata/logic_test/scale#L61)):*
> ```
>   b INT2,
>   UNIQUE INDEX a (b)
> )
> ```

Can you add a test that demonstrates that `SHOW CREATE TABLE` remembers the column types correctly? That test should probably go in a different test file, like `table`.

---

*[pkg/sql/logictest/testdata/logic_test/scale, line 76 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnjzN4ea9IvNQn6Su8H:-KnjzN4ea9IvNQn6Su8I:b-bb2s1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/logictest/testdata/logic_test/scale#L76)):*
> ```
> 
> statement error integer out of range for type INT2 \(column "b"\)
> INSERT INTO tc VALUES (32768)
> ```

We might want to consider adding a test case like `INSERT INTO tc VALUES (60000-59999)` with a comment to demonstrate that these column type bounds are only enforced once inserted - in case someone wants to change that one day

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-Knk0fPvpDT9NtIx66Lw:b-siq3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
> ```protobuf
>     INT2 = 3;
>     INT4 = 4;
>     SMALLINT = 5;
> ```

I think we don't want `SMALLINT` here, right? Since it's equivalent to `INT4`? I believe for Postgres compatibility we want just `INT4`.

---

*[pkg/sql/sqlbase/table.go, line 1629 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-8PqPVMTC7LEmtVA:-Knk-8PqPVMTC7LEmtVB:bx9di9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1629)):*
> ```Go
> 				//to attempt to store shorter or longer bit strings. bit varying
> 				//data is of variable length up to the maximum length n; longer
> 				//strings will be rejected." Bits are unsigend, so we need to
> ```

s/unsigend/unsigned/

---

*[pkg/sql/sqlbase/table.go, line 1636 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-BwcLpv-O_A17uOn:-Knk-BwdjR1iHCoqBYQq:bxfkda3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1636)):*
> ```Go
> 				}
> 
> 				// We're performing bounds checks inline with Go's implementation of min and max ints in Math.go.
> ```

nit: comment is too long - while you're in here could you please make it less than 80 characters? thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-Knk1gx-WR-q8l4fa6Pr:b-832rmg)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Being tracked on https://github.com/cockroachdb/cockroach/issues/16769, but will split into another PR.

---

Review status: 10 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/scale, line 61 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk2RUEKccrcHEscNJd:-KnkCPyRSy0ZzzURFWET:b-54wavk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/logictest/testdata/logic_test/scale#L61)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can you add a test that demonstrates that `SHOW CREATE TABLE` remembers the column types correctly? That test should probably go in a different test file, like `table`.
</blockquote></details>

That's already in logictest/table. It's a bit hard to see because I refactored it a little.

---

*[pkg/sql/logictest/testdata/logic_test/scale, line 76 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnjzN4ea9IvNQn6Su8H:-KnkXCgmMeyUkRjGNtf6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/logictest/testdata/logic_test/scale#L76)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We might want to consider adding a test case like `INSERT INTO tc VALUES (60000-59999)` with a comment to demonstrate that these column type bounds are only enforced once inserted - in case someone wants to change that one day
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-KnkCkNWXMZ1qsp8jgcL:b-ugj5rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think we don't want `SMALLINT` here, right? Since it's equivalent to `INT4`? I believe for Postgres compatibility we want just `INT4`.
</blockquote></details>

SMALLINT is an alias for INT4 in Postgres, so not having it here would break the show create table behavior.

---

*[pkg/sql/sqlbase/table.go, line 1629 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-8PqPVMTC7LEmtVA:-KnkDBE5uKqurGOi0i_X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1629)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/unsigend/unsigned/
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1636 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-BwcLpv-O_A17uOn:-KnkDC7ykJo7fDrqmuMy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1636)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: comment is too long - while you're in here could you please make it less than 80 characters? thanks!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KnkXGqfNG0mypso1t1n:b7bbjp)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 10 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/scale, line 61 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk2RUEKccrcHEscNJd:-Knkej13aNcw2Ogjc8oo:b7aaikl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/logictest/testdata/logic_test/scale#L61)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

That's already in logictest/table. It's a bit hard to see because I refactored it a little.
</blockquote></details>

D'oh! Thanks.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-Knkf0r8A4YzA70z78Oa:bbwanuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

SMALLINT is an alias for INT4 in Postgres, so not having it here would break the show create table behavior.
</blockquote></details>

Right - but actually we're okay with `SHOW CREATE TABLE` returning those aliases. Postgres always reports `smallint` as the "sql type name" of an int4 column, regardless of how it was created. We can do the same thing, but with `int4` as the canonical name.

Furthermore, only adding `SMALLINT` here is inconsistent, as if we wanted to go down this road we'd also have to add e.g. `BIGINT` and `DOUBLE PRECISION`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 10 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-Knl63vNIDjVLa7ckuB_:bshpd52) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Right - but actually we're okay with `SHOW CREATE TABLE` returning those aliases. Postgres always reports `smallint` as the "sql type name" of an int4 column, regardless of how it was created. We can do the same thing, but with `int4` as the canonical name.

Furthermore, only adding `SMALLINT` here is inconsistent, as if we wanted to go down this road we'd also have to add e.g. `BIGINT` and `DOUBLE PRECISION`.
</blockquote></details>

1) "smallint" seems better as a canonical name, not because pg does it but because it's closer to standard SQL

2) yes I would argue real, bigint and double precision should land here eventually too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 10 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-Knl7-dEE1MURNr1BEds:br9ugaf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) "smallint" seems better as a canonical name, not because pg does it but because it's closer to standard SQL

2) yes I would argue real, bigint and double precision should land here eventually too
</blockquote></details>

1. Agreed
2. The names of these enum values don't have to be the canonical names. But we shouldn't have two enum values that map to the same thing - that doesn't make sense to me! There are 3 ways of spelling `int8` (`bigint`, `int8`, `int64`), but there should be exactly 1 `VisibleType` enum value for `int8`. This matches how Postgres works - you can spell a type many different ways, but that type always gets normalized into a single spelling for its equivalent of `SHOW CREATE`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 10 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-KnlNSyv381gPk1WJEPb:b9l1cv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

1. Agreed
2. The names of these enum values don't have to be the canonical names. But we shouldn't have two enum values that map to the same thing - that doesn't make sense to me! There are 3 ways of spelling `int8` (`bigint`, `int8`, `int64`), but there should be exactly 1 `VisibleType` enum value for `int8`. This matches how Postgres works - you can spell a type many different ways, but that type always gets normalized into a single spelling for its equivalent of `SHOW CREATE`.
</blockquote></details>

aw, ok thanks for explaining. Aye, good to choose a single canonical name. I'm in favor of the standard sql names: bigint/smallint

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 3 of 15 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/structured.proto, line 74 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk0fPvpDT9NtIx66Lv:-KnofveKa01zbzICB59W:bgdfm6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/structured.proto#L74)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

aw, ok thanks for explaining. Aye, good to choose a single canonical name. I'm in favor of the standard sql names: bigint/smallint
</blockquote></details>

I added a mapping from type name to canonical type in table.go, I can move it if either of you think it should go somewhere else.  Int is not the postgres canonical type for int, but I left it as such because a lot of our own code uses that. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

knz commented:
I'm comfortable with the result here, but will defer to Jordan for a green light.

---

Reviewed 12 of 12 files at r8.
Review status: 14 of 15 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 1138 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnpfqFwel8tCOUQwhJP:-KnpfqFxF9uNsDyaL6nz:b-bz4zhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdebbef52bd60bb1d48f0998cf6eb69a92133d0f/pkg/sql/logictest/testdata/logic_test/pg_catalog#L1138)):*
> ```
> 0
> 
> #statement ok
> ```

Why are these commented out? If it's intentional, add a comment above to explain why they are commented out.

---

*[pkg/sql/sqlbase/table.go, line 1639 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnpgFjCJJEQj9NBEw8D:-KnpgFjCJJEQj9NBEw8E:b-m0d1qv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdebbef52bd60bb1d48f0998cf6eb69a92133d0f/pkg/sql/sqlbase/table.go#L1639)):*
> ```Go
> 
> 				//https://www.postgresql.org/docs/9.5/static/datatype-bit.html
> 				//"bit type data must match the length n exactly; it is an error
> ```

nit: we usually separate the comment prefix from the text with a space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KnpgKxlmaP6rm3y6mrq:b-az2f67)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the little typo and nit comments left!

---

Review status: 14 of 15 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 1138 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnpfqFwel8tCOUQwhJP:-KnqOvIyuHAqtkHXC8lB:bel1h56) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdebbef52bd60bb1d48f0998cf6eb69a92133d0f/pkg/sql/logictest/testdata/logic_test/pg_catalog#L1138)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why are these commented out? If it's intentional, add a comment above to explain why they are commented out.
</blockquote></details>

Yeah this should just have above it: ## TODO(masha): #16769 

---

*[pkg/sql/sqlbase/table.go, line 1629 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-8PqPVMTC7LEmtVA:-KnqPYSTtHr8dC0THGWW:bo0n6cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1629)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hmm maybe you didn't repush these latest changes? I still see the typo and the long line unless reviewable is malfunctioning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KnqRBGb71j7FIzUr1Rn:b-roblp0)*
<!-- Sent from Reviewable.io -->

knz commented:
something weird happened with your branch, I'm not sure you want to see the merge commits in there?

m-schneider commented:
Rebase gone badly, I'm pretty sure I fixed it now.

---

Review status: 11 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 1138 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnpfqFwel8tCOUQwhJP:-Knu8UY7s5-K0DMD5uRI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdebbef52bd60bb1d48f0998cf6eb69a92133d0f/pkg/sql/logictest/testdata/logic_test/pg_catalog#L1138)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yeah this should just have above it: ## TODO(masha): #16769 
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1629 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-Knk-8PqPVMTC7LEmtVA:-Knu91SMTkTUSD4hgJsk:b-fsdl1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/a46cb814b73865a75f4ce6ab38f75e378f262bd8/pkg/sql/sqlbase/table.go#L1629)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm maybe you didn't repush these latest changes? I still see the typo and the long line unless reviewable is malfunctioning.
</blockquote></details>

I must have reverted something on my end. Fixed now!

---

*[pkg/sql/sqlbase/table.go, line 1639 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-KnpgFjCJJEQj9NBEw8D:-Knu8z8TvTDMXzdaRqba:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdebbef52bd60bb1d48f0998cf6eb69a92133d0f/pkg/sql/sqlbase/table.go#L1639)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: we usually separate the comment prefix from the text with a space.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KnuK5DaOyexPVDDDaSn:b-nlmhs4)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 4 of 4 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :tada: 

---

Reviewed 1 of 12 files at r1, 2 of 4 files at r7, 8 of 12 files at r8, 4 of 4 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16720#-:-KnuUx0KGboKqS2wLb8Y:by6fmpn)*
<!-- Sent from Reviewable.io -->
